File: RentTracker/Models/Enums.swift
import Foundation

// MARK: - Property Type
enum PropertyType: String, CaseIterable, Identifiable {
    case residential = "residential"
    case commercial = "commercial"
    
    var id: String { self.rawValue }
    
    var localizedName: String {
        switch self {
        case .residential:
            return NSLocalizedString("Residential", comment: "Residential property type")
        case .commercial:
            return NSLocalizedString("Commercial", comment: "Commercial property type")
        }
    }
}

// MARK: - Payment Cycle
enum PaymentCycle: String, CaseIterable, Identifiable {
    case monthly = "monthly"
    case bimonthly = "bimonthly"
    case quarterly = "quarterly"
    case yearly = "yearly"
    
    var id: String { self.rawValue }
    
    var localizedName: String {
        switch self {
        case .monthly:
            return NSLocalizedString("Monthly", comment: "Monthly payment cycle")
        case .bimonthly:
            return NSLocalizedString("Bi-monthly", comment: "Bi-monthly payment cycle")
        case .quarterly:
            return NSLocalizedString("Quarterly", comment: "Quarterly payment cycle")
        case .yearly:
            return NSLocalizedString("Yearly", comment: "Yearly payment cycle")
        }
    }
    
    var monthsInterval: Int {
        switch self {
        case .monthly: return 1
        case .bimonthly: return 2
        case .quarterly: return 3
        case .yearly: return 12
        }
    }
}

// MARK: - Payment Type
enum PaymentType: String, CaseIterable, Identifiable {
    case rent = "rent"
    case lateFee = "lateFee"
    case deposit = "deposit"
    case depositReturn = "depositReturn"
    
    var id: String { self.rawValue }
    
    var localizedName: String {
        switch self {
        case .rent:
            return NSLocalizedString("Rent", comment: "Rent payment type")
        case .lateFee:
            return NSLocalizedString("Late Fee", comment: "Late fee payment type")
        case .deposit:
            return NSLocalizedString("Deposit", comment: "Deposit payment type")
        case .depositReturn:
            return NSLocalizedString("Deposit Return", comment: "Deposit return payment type")
        }
    }
    
    var isIncome: Bool {
        switch self {
        case .rent, .lateFee, .deposit:
            return true
        case .depositReturn:
            return false
        }
    }
}

// MARK: - Payment Method
enum PaymentMethod: String, CaseIterable, Identifiable {
    case bankTransfer = "bankTransfer"
    case wechatPay = "wechatPay"
    case cash = "cash"
    
    var id: String { self.rawValue }
    
    var localizedName: String {
        switch self {
        case .bankTransfer:
            return NSLocalizedString("Bank Transfer", comment: "Bank transfer payment method")
        case .wechatPay:
            return NSLocalizedString("WeChat Pay", comment: "WeChat Pay payment method")
        case .cash:
            return NSLocalizedString("Cash", comment: "Cash payment method")
        }
    }
    
    var iconName: String {
        switch self {
        case .bankTransfer: return "building.columns"
        case .wechatPay: return "message"
        case .cash: return "banknote"
        }
    }
}

// MARK: - Expense Category
enum ExpenseCategory: String, CaseIterable, Identifiable {
    case maintenance = "maintenance"
    case repair = "repair"
    case other = "other"
    
    var id: String { self.rawValue }
    
    var localizedName: String {
        switch self {
        case .maintenance:
            return NSLocalizedString("Maintenance", comment: "Maintenance expense category")
        case .repair:
            return NSLocalizedString("Repair", comment: "Repair expense category")
        case .other:
            return NSLocalizedString("Other", comment: "Other expense category")
        }
    }
    
    var iconName: String {
        switch self {
        case .maintenance: return "wrench"
        case .repair: return "hammer"
        case .other: return "ellipsis.circle"
        }
    }
}

File: app/lib/core/enums/enums.dart
/// Core enumerations for the RentTracker app
///
/// These enums define the valid values for various domain concepts
/// and must match the database schema constraints.
library;

// ==============================================================================
// PROPERTY TYPE
// ==============================================================================

/// Type of rental property
enum PropertyType {
  /// Residential property (apartments, houses, etc.)
  residential('residential'),

  /// Commercial property (offices, retail spaces, etc.)
  commercial('commercial');

  /// String value for database storage
  final String value;

  const PropertyType(this.value);

  /// Convert string to PropertyType enum
  ///
  /// Throws [ArgumentError] if the string doesn't match any enum value
  static PropertyType fromString(String value) {
    return PropertyType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => throw ArgumentError(
        'Invalid PropertyType: $value. Valid values are: ${PropertyType.values.map((e) => e.value).join(", ")}',
      ),
    );
  }
}

// ==============================================================================
// PAYMENT CYCLE
// ==============================================================================

/// Frequency of rent payments
enum PaymentCycle {
  /// Monthly payments (every 1 month)
  monthly('monthly', 1),

  /// Bimonthly payments (every 2 months)
  bimonthly('bimonthly', 2),

  /// Quarterly payments (every 3 months)
  quarterly('quarterly', 3),

  /// Yearly payments (every 12 months)
  yearly('yearly', 12);

  /// String value for database storage
  final String value;

  /// Number of months between payments (used for schedule generation)
  final int months;

  const PaymentCycle(this.value, this.months);

  /// Convert string to PaymentCycle enum
  ///
  /// Throws [ArgumentError] if the string doesn't match any enum value
  static PaymentCycle fromString(String value) {
    return PaymentCycle.values.firstWhere(
      (cycle) => cycle.value == value,
      orElse: () => throw ArgumentError(
        'Invalid PaymentCycle: $value. Valid values are: ${PaymentCycle.values.map((e) => e.value).join(", ")}',
      ),
    );
  }
}

// ==============================================================================
// PAYMENT TYPE
// ==============================================================================

/// Type of payment transaction
enum PaymentType {
  /// Regular rent payment
  rent('rent'),

  /// Late fee payment (penalty for overdue rent)
  lateFee('lateFee'),

  /// Security deposit payment (paid at contract start)
  deposit('deposit'),

  /// Security deposit return (refunded at contract end)
  depositReturn('depositReturn');

  /// String value for database storage
  final String value;

  const PaymentType(this.value);

  /// Convert string to PaymentType enum
  ///
  /// Throws [ArgumentError] if the string doesn't match any enum value
  static PaymentType fromString(String value) {
    return PaymentType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => throw ArgumentError(
        'Invalid PaymentType: $value. Valid values are: ${PaymentType.values.map((e) => e.value).join(", ")}',
      ),
    );
  }
}

// ==============================================================================
// PAYMENT METHOD
// ==============================================================================

/// Method used for payment
enum PaymentMethod {
  /// Bank transfer (wire transfer, ACH, etc.)
  bankTransfer('bankTransfer'),

  /// WeChat Pay (popular in China)
  wechatPay('wechatPay'),

  /// Cash payment
  cash('cash');

  /// String value for database storage
  final String value;

  const PaymentMethod(this.value);

  /// Convert string to PaymentMethod enum
  ///
  /// Throws [ArgumentError] if the string doesn't match any enum value
  static PaymentMethod fromString(String value) {
    return PaymentMethod.values.firstWhere(
      (method) => method.value == value,
      orElse: () => throw ArgumentError(
        'Invalid PaymentMethod: $value. Valid values are: ${PaymentMethod.values.map((e) => e.value).join(", ")}',
      ),
    );
  }
}

// ==============================================================================
// EXPENSE CATEGORY
// ==============================================================================

/// Category of property expense
enum ExpenseCategory {
  /// Regular maintenance (cleaning, landscaping, etc.)
  maintenance('maintenance'),

  /// Repairs (fixing broken items)
  repair('repair'),

  /// Other miscellaneous expenses
  other('other');

  /// String value for database storage
  final String value;

  const ExpenseCategory(this.value);

  /// Convert string to ExpenseCategory enum
  ///
  /// Throws [ArgumentError] if the string doesn't match any enum value
  static ExpenseCategory fromString(String value) {
    return ExpenseCategory.values.firstWhere(
      (category) => category.value == value,
      orElse: () => throw ArgumentError(
        'Invalid ExpenseCategory: $value. Valid values are: ${ExpenseCategory.values.map((e) => e.value).join(", ")}',
      ),
    );
  }
}
